search/beam.hpp

```diff
@@ -0,0 +1,213 @@
+  typedef typename D::Oper Oper;
+
+  struct Node {
+    int openind;
+    Node *parent;
+    PackedState state;
+    Oper op, pop;
+    Cost f, g;
+
+    Node() : openind(-1) {
+    }
+
+    static ClosedEntry<Node, D> &closedentry(Node *n) {
+      return n->closedent;
+    }
+
+    static PackedState &key(Node *n) {
+      return n->state;
+    }
+
+    /* Set index of node on open list. */
+    static void setind(Node *n, int i) {
+      n->openind = i;
+    }
+
+    /* Get index of node on open list. */
+    static int getind(const Node *n) {
+      return n->openind;
+    }
+
+    /* Indicates whether Node a has better value than Node b. */
+    static bool pred(Node *a, Node *b) {
+      if (a->f == b->f)
+        return a->g > b->g;
+      return a->f < b->f;
+    }
+
+    /* Priority of node. */
+    static Cost prio(Node *n) {
+      return n->f;
+    }
+
+    /* Priority for tie breaking. */
+    static Cost tieprio(Node *n) {
+      return n->g;
+    }
+
+    private:
+		ClosedEntry<Node, D> closedent;
+    
+  };
+
+  BeamSearch(int argc, const char *argv[]) :
+    SearchAlgorithm<D>(argc, argv), closed(30000001) {
+    dropdups = false;
+    for (int i = 0; i < argc; i++) {
+      if (i < argc - 1 && strcmp(argv[i], "-width") == 0)
+        width = atoi(argv[++i]);
+      if (strcmp(argv[i], "-dropdups") == 0)
+        dropdups = true;
+    }
+
+    if (width < 1)
+      fatal("Must specify a >0 beam width using -width");
+    
+    nodes = new Pool<Node>();
+  }
+
+  ~BeamSearch() {
+    delete nodes;
+  }
+
+  void search(D &d, typename D::State &s0) {
+    this->start();
+    closed.init(d);
+
+    Node *n0 = init(d, s0);
+    closed.add(n0);
+    open.push(n0);
+
+    int depth = 0;
+
+
+    bool solved = false;
+    
+    while (!open.empty() && !solved && !SearchAlgorithm<D>::limit()) {
+      depth++;
+      
+      Node **beam = new Node*[width];
+      int c;
+      for(c = 0; c < width && !open.empty(); c++) {
+        Node *n = open.pop();
+
+        beam[c] = n;
+      }
+
+      open.clear();
+      
+      for(int i = 0; i < c && !solved && !SearchAlgorithm<D>::limit(); i++) {
+        Node *n = beam[i];
+        State buf, &state = d.unpack(buf, n->state);
+
+        if (d.isgoal(state)) {
+          solpath<D, Node>(d, n, this->res);
+          solved = true;
+          break;
+        }
+
+        expand(d, n, state);
+      }
+
+      delete[] beam;
+    }
+    this->finish();
+  }
+
+  virtual void reset() {
+    SearchAlgorithm<D>::reset();
+    open.clear();
+    closed.clear();
+    delete nodes;
+    nodes = new Pool<Node>();
+  }
+
+  virtual void output(FILE *out) {
+    SearchAlgorithm<D>::output(out);
+    closed.prstats(stdout, "closed ");
+    dfpair(stdout, "open list type", "%s", open.kind());
+    dfpair(stdout, "node size", "%u", sizeof(Node));
+  }
+
+
+private:
+
+	void expand(D &d, Node *n, State &state) {
+		SearchAlgorithm<D>::res.expd++;
```

Comment:
the indents changed here from spaces before to tabs. i don't recall what the other files in the repo use, but prefer sticking to that